File: backend/app/schemas/staff.py
from pydantic import BaseModel, Field, EmailStr
from typing import Optional, List, Dict, Any, Union
from datetime import datetime, date
from .staff_position import StaffPosition
from .user import User

class StaffPositionBase(BaseModel):
    position_id: int
    is_primary: bool = False


class StaffPositionCreate(StaffPositionBase):
    pass


class StaffPosition(StaffPositionBase):
    id: int
    staff_id: int
    position_name: Optional[str] = None
    
    class Config:
        orm_mode = True


class StaffBase(BaseModel):
    """Базовая схема сотрудника"""
    first_name: str
    last_name: str
    middle_name: Optional[str] = None
    email: Optional[EmailStr] = None
    phone: Optional[str] = None
    hire_date: Optional[date] = None
    organization_id: Optional[int] = None
    user_id: Optional[int] = None
    photo_path: Optional[str] = None
    document_paths: Optional[Dict[str, str]] = None
    is_active: bool = True
    
class StaffCreate(StaffBase):
    """Схема для создания сотрудника"""
    positions: Optional[List[StaffPositionCreate]] = None
    create_user: bool = False # Флаг для создания связанного пользователя
    password: Optional[str] = None
    
class StaffUpdate(StaffBase):
    """Схема для обновления сотрудника"""
    first_name: Optional[str] = None
    last_name: Optional[str] = None
    middle_name: Optional[str] = None
    email: Optional[EmailStr] = None
    phone: Optional[str] = None
    hire_date: Optional[date] = None
    user_id: Optional[int] = None
    photo_path: Optional[str] = None
    document_paths: Optional[Dict[str, str]] = None
    is_active: Optional[bool] = None
    create_user: Optional[bool] = None  # Флаг для создания связанного пользователя при обновлении
    positions: Optional[List[StaffPositionCreate]] = None
    password: Optional[str] = None
    
class StaffInDBBase(StaffBase):
    """Базовая схема для сотрудника в БД"""
    id: int
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True
        
class Staff(StaffInDBBase):
    """Схема для возврата сотрудника через API"""
    positions: Optional[List[StaffPosition]] = None  # Связанные должности
    user: Optional[User] = None  # Связанный пользователь
    organization_name: Optional[str] = None  # Название организации

# Новая схема для ответа при создании сотрудника с пользователем
class StaffCreateResponse(Staff):
    activation_code: Optional[str] = None 

File: frontend/src/types/staff.ts
import { User } from './user';
import { Position } from './position';

// Интерфейс, описывающий данные должности, возвращаемые API в массиве positions
export interface StaffPositionInfo {
  id: number;
  staff_id: number;
  position_id: number;
  is_primary: boolean;
  position_name: string | null; // Используем это поле
  // start_date, end_date, created_at, updated_at можно добавить, если они нужны на фронте
}

export interface Staff {
  id: number;
  first_name: string;
  last_name: string;
  middle_name?: string | null;
  email?: string | null;
  phone?: string | null;
  hire_date?: string | null; // ISO формат даты
  user_id?: number | null;
  organization_id?: number | null; // Добавляем organization_id
  photo_path?: string | null;
  document_paths?: Record<string, string> | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
  user?: User | null; // Пользователь может быть null
  positions?: StaffPositionInfo[]; // Меняем staff_positions на positions и используем новый тип
  organization_name?: string | null; 
}

// Интерфейс для создания должности при создании/обновлении сотрудника
export interface StaffPositionCreateData {
  position_id: number;
  is_primary: boolean;
}

export interface StaffCreate {
  first_name: string;
  last_name: string;
  middle_name?: string;
  email?: string;
  phone?: string;
  hire_date?: string;
  organization_id?: number;
  is_active?: boolean;
  positions?: StaffPositionCreateData[]; // Используем массив должностей
  create_user?: boolean; 
  password?: string; // Пароль опционален, только если create_user=true
}

export interface StaffCreateResponse extends Staff {
  activation_code?: string | null;
}

export interface StaffUpdate {
  first_name?: string;
  last_name?: string;
  middle_name?: string;
  email?: string;
  phone?: string;
  hire_date?: string;
  organization_id?: number; // Можно обновлять организацию
  is_active?: boolean;
  positions?: StaffPositionCreateData[]; // Можно обновлять должности
  // user_id, create_user, password обычно не обновляются этим методом
  // Если нужно обновить user_id или создать/изменить пользователя, нужны отдельные поля/логика
}

// Интерфейс для RocketChat данных сотрудника
export interface StaffRocketChat {
  username: string; // Обычно email или другой уникальный идентификатор
  email: string;
  name: string; // Полное имя
  password?: string; // Опционально, для создания
  roles?: string[]; // Роли в Rocket.Chat
  active?: boolean;
} 